servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java

```diff
@@ -130,57 +125,77 @@ public void onSubscribe(final Cancellable cancellable) {
                 }
-            } else if (newScheme == null && scheme != null) {
-                // Rewrite relative-form location to absolute-form request-target for multi-address client
-                newRequest.requestTarget(scheme + "://" + newRequest.headers().get(HOST) + newRequest.requestTarget());
-            }
-            if (LOGGER.isTraceEnabled()) {
-                LOGGER.trace("Execute redirect to '{}' for original request '{}'",
-                        result.headers().get(LOCATION), redirectSingle.originalRequest);
+
+                StreamingHttpRequest newRequest = prepareRedirectRequest(request, response,
+                        redirectSingle.requester);
+                if (newRequest == null) {
+                    terminalDelivered = true;
+                    target.onSuccess(response);
+                    return;
+                }
+
+                final String newScheme = newRequest.scheme();
+                final boolean relative = isRelative(request, scheme, newRequest);
+                if (!relative && !redirectSingle.allowNonRelativeRedirects) {
+                    LOGGER.debug(
+                        "Ignoring non-relative redirect to '{}' for request '{}': Only relative redirects are allowed",
+                        newRequest.requestTarget(), request);
+                    terminalDelivered = true;
+                    target.onSuccess(response);
+                    return;
+                }
+
+                if (!redirectSingle.config.shouldRedirectPredicate().test(relative, redirectCount, request, response)) {
+                    terminalDelivered = true;
+                    target.onSuccess(response);
+                    return;
+                }
+
+                if (relative) {
+                    if (redirectSingle.allowNonRelativeRedirects && newScheme == null && scheme != null) {
+                        // Rewrite origin-form location to absolute-form request-target for multi-address client:
+                        newRequest.requestTarget(scheme + "://" + newRequest.headers().get(HOST) +
+                                newRequest.requestTarget());
+                    }
+                    if (!redirectSingle.allowNonRelativeRedirects && newScheme != null) {
+                        // Rewrite absolute-form location to origin-form request-target in case only relative redirects
+                        // are supported:
+                        newRequest.requestTarget(
+                                absoluteToRelativeFormRequestTarget(newRequest.requestTarget(), newScheme));
+                    }
+                }
+
+                newRequest = redirectSingle.config.redirectRequestTransformer()
+                        .apply(relative, request, response, newRequest);
+
+                if (LOGGER.isTraceEnabled()) {
+                    LOGGER.trace("Executing redirect to '{}' for request '{}'", response.headers().get(LOCATION),
```

Comment:
nit: if shouldRedirect to return the LOCATION as suggested in https://github.com/apple/servicetalk/pull/1792/files#r707906058 you could also use it here too.